docs/modules/ROOT/pages/references/style-guide.adoc

```diff
@@ -0,0 +1,416 @@
+  Optional parameters should have a default, or `null` if a sentinel value is needed.
++
+[source,jsonnet]
+----
+local newAnimal(name, age, isCat = true) = { ... }
+----
+
+* Wrap parameter declarations by putting one parameter per line with 2 extra spaces of indentation, to differentiate from the function body.
+  Doing this is always acceptable, even if the definition would not wrap.
++
+[source,jsonnet]
+----
+local newAnimal(
+    name,
+    age,
+    isCat = true) = {
+  name: name,
+  …
+}
+----
+
+== Defining functions
+
+* Don't define functions within objects.
+  Such objects will fail to render.
+  The exception to this rule is the last object within a library file.
+* Functions which return single values (rather than an object) should use parentheses `()` to enclose their bodies if they're multi-line, identically to how braces would be used.
++
+[source,jsonnet]
+----
+{
+  multiply(number1, number2):
+    (
+      number1 * number 2
+    ),
+}
+----
+
+=== Using libraries
+
+* Import all dependencies at the top of the file and given them names related to the imported file itself.
+  This makes it easy to see what other files you depend on as the file grows.
++
+[source,jsonnet]
+----
+// CORRECT
+local animal = import "animal.libsonnet";
+animal.newAnimal("Finnegan", 3);
+
+// AVOID
+(import "animal.libsonnet").newAnimal("Finnegan, 3);
+----
+
+* Keep function parameters on a single line or put one parameter per line when calling functions.
+* When putting one parameter per line for a function call, add a line break (`\n`) after the opening `(`.
++
+[source,jsonnet]
+----
+// CORRECT
+animal.newAnimal("Finnegan", 3);
+animal.newAnimal(
+  name = "Finnegan",
+  age = 3,
+);
+animal.newAnimal(
+  "Finnegan",
+  3,
+);
+
+// INCORRECT
+animal.newAnimal("Finnegan",
+  3,
+  42,
+);
+----
+
+=== File structure
+
+* Jsonnet files which can be materialized with no further inputs should end with the `.jsonnet` suffix.
+* Jsonnet files which require parameters to be materialized or which are libraries should end with the `.libjsonnet` suffix.
+* Files in `lib` always are libraries which should never be materialized and must be named accordingly.
+  Those files are considered part of a public API.
+  Treat functions in libraries accordingly and look out for breaking changes.
+
+==== Documentation style
+
+* Use `//` for inline comments.
+* Use https://www.doxygen.nl/manual/docblocks.html[Docblocks] to document functions.
++
+[source,jsonnet]
+----
+/**
+ * Multicellular, eukaryotic organism of the kingdom Animalia
+ *
+ * @param name Name by which this animal may be called.
+ * @param age Number of years (rounded to nearest int) animal has been alive.
+ * @returns an object describing the animal.
+ */
+local Animal(name, age) = { … }
+----
+
+* Always put documentation at the top of each Jsonnet file or library to indicate its purpose.
```

Comment:
```suggestion
* Put a  https://www.doxygen.nl/manual/docblocks.html[Docblock] at the top of each Jsonnet file or library to indicate its purpose.
```

And an example (might be helpful to get an idea of what we expect from top-level documentation):

```
/**
  * \file widgetpatch.libjsonnet
  *
  * The widget patching library provides functions to adjust the content of prepackaged widgets.
  *
  * Generally, functions will return a widget patch which can be merged with a Widget to make the corresponding adjustment.
  *
  * For example, you can configure the widget's owner with `AddOwner`:
  *
  * \code
  * local wpatch = import "widgetpatch.libjsonnet";
  *
  * local thewidget = inventory.widget_a + wpatch.AddOwner('Synthesizer');
  * \endcode
  *
  * See the individual function documentation for detailed explanation of each function's behavior.
  */